### DISCUSSION TITLE
Periodic 3D Delaunay mesh in 1-sheeted covering

### DISCUSSION THREAD

**Discussion by gy114** [CGAL/cgal#8063] [Q&A] [answered]
Hi, 

I need to create a 3D periodic Delaunay mesh incrementally. This has to be done in an incremental way because I need to have the vertex handle, from which I can assign the info(). 

As the number of points increases, I realized that the mesh is automatically converted to 1-sheeted covering at some point. This should not be an issue, until I had to access the geometry of those "unique cells" near the boundary of the cuboid. These unique cells should have vertices that are outside the cuboid, which are actually periodic images of some points "rounded to the other side" of the cuboid. However, in 1-sheeted covering, the offset for all points are (0,0,0), without memory of their offsets in those unique cells near the boundary. These cells are therefore "stretched" and "intersecting" with other canonical cells, which is logically correct but geometrically strange. 

I can convert to 27-sheeted covering and resolve this issue. But that costs a lot of memory. 
In 1-sheeted covering, is there any way to get the offset information for those unique cells near the boundary? 

Thanks.

**Comment by MaelRL** [comment, marked as answer]
There are two types of offsets, which you have to distinguish between:
- offsets of vertices: as you write, when you are in 27 sheet, vertices are duplicate over 27 copies, and 26 copies are offsets of the canonical one. This offset lives within the vertex base class.
- offsets of vertices **within a cell**: this is the offset that you need to apply (on top of the vertex offset above) to get the correct geometry of the cell. There is one offset per vertex, and it's usually ((0,0,0); (0,0,0); (0,0,0); (0,0,0)) since most cells do not cross over the boundaries. This offset lives in the cell base class.

You could compute the geometric position manually, getting vh->offset() and cell->offset(int) and combining properly, but there are functions in the the triangulation class to do that for you:
- If you use P3T3::point(vh), you get the position of the vertex.
- If you use the function P3T3::point(ch, i), you get the position of the i-th vertex of ch and these 4 positions will be consistent with one another as to give you the proper geometry of the cell.

Note that unique iterators work even in 1-sheet.

**Reply by gy114** [reply]
Worked perfectly. Many thanks!
